File: app/src/main/java/com/corphish/nightlight/ProfilesActivity.java
package com.corphish.nightlight;

import android.os.Bundle;
import android.support.design.widget.BottomSheetDialog;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatButton;
import android.support.v7.widget.AppCompatEditText;
import android.support.v7.widget.AppCompatSeekBar;
import android.support.v7.widget.AppCompatSpinner;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.TextView;

import com.corphish.nightlight.data.Constants;
import com.corphish.nightlight.engine.ProfilesManager;
import com.corphish.nightlight.helpers.PreferenceHelper;

import java.util.ArrayList;
import java.util.List;

public class ProfilesActivity extends AppCompatActivity implements ProfilesManager.DataChangeListener{

    private View creatorView;
    private AppCompatEditText editText;
    private TextView editTextError, settingTitle1, settingTitle2, cancel;
    private AppCompatSpinner modes;
    private AppCompatSeekBar settingParam1, settingParam2;
    private AppCompatButton ok;
    private SwitchCompat nlSwitch;
    private BottomSheetDialog bottomSheetDialog, optionsDialog;
    private View optionsView;

    private int currentModeSelection = Constants.NL_SETTING_MODE_FILTER;
    private ProfilesManager.Profile curProfile = null;

    private final int MODE_CREATE   =   0;
    private final int MODE_EDIT     =   1;
    private int curMode             =   MODE_CREATE;

    private ProfilesManager profilesManager;

    private ProfilesAdapter profilesAdapter;

    private ArrayList<ProfilesManager.Profile> profiles;

    private View emptyView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_profiles);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) actionBar.setDisplayHomeAsUpEnabled(true);

        findViewById(R.id.fab).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                curProfile = null;
                curMode = MODE_CREATE;
                bottomSheetDialog = new BottomSheetDialog(ProfilesActivity.this, R.style.BottomSheetDialogDark);
                initProfileCreatorViews();
                bottomSheetDialog.setContentView(creatorView);
                bottomSheetDialog.setCancelable(false);
                bottomSheetDialog.show();
            }
        });

        emptyView = findViewById(R.id.emptyView);

        initProfilesManager();
        initViews();
    }

    private void initProfilesManager() {
        profilesManager = new ProfilesManager(this);
        profilesManager.registerDataChangeListener(this);
        profilesManager.loadProfiles();
        profiles = profilesManager.getProfilesList();
    }

    private void initViews() {
        RecyclerView recyclerView = findViewById(R.id.profiles_holder);
        profilesAdapter = new ProfilesAdapter();
        profilesAdapter.setProfiles(profiles);

        recyclerView.invalidateItemDecorations();
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(profilesAdapter);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setNestedScrollingEnabled(false);
        recyclerView.setHasFixedSize(false);

        profilesAdapter.notifyDataSetChanged();
    }

    @Override
    public void onDataChanged(int newDataSize) {
        if (newDataSize < 1) emptyView.setVisibility(View.VISIBLE);
        else emptyView.setVisibility(View.GONE);
    }

    private class ProfilesAdapter extends RecyclerView.Adapter<ProfilesActivity.ProfilesAdapter.CustomViewHolder> {
        private List<ProfilesManager.Profile> profiles;

        void setProfiles(List<ProfilesManager.Profile> profiles) {
            this.profiles = profiles;
        }

        public class CustomViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
            final TextView name, desc;
            CustomViewHolder(View v) {
                super(v);

                name = v.findViewById(R.id.profile_name);
                desc = v.findViewById(R.id.profile_desc);

                v.setOnClickListener(this);
            }

            @Override
            public void onClick(View v) {
                curProfile = profiles.get(getAdapterPosition());
                optionsDialog = new BottomSheetDialog(ProfilesActivity.this, R.style.BottomSheetDialogDark);
                getOptionsView();
                optionsDialog.setContentView(optionsView);
                optionsDialog.show();
            }
        }



        @Override
        public CustomViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View itemView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.card_profile_item, parent, false);
            return new CustomViewHolder(itemView);
        }

        @Override
        public void onBindViewHolder(final CustomViewHolder holder, int position) {
            holder.name.setText(profiles.get(position).getName());
            holder.desc.setText(getDescription(profiles.get(position)));
        }

        @Override
        public int getItemCount() {
            return profiles.size();
        }

        private String getDescription(ProfilesManager.Profile profile) {
            String desc = "";
            desc += getString(R.string.app_name) + " : " +
                    (profile.isSettingEnabled() ? getString(R.string.on).toLowerCase() : getString(R.string.off).toLowerCase()) + ", ";
            if (profile.getSettingMode() == Constants.NL_SETTING_MODE_TEMP) {
                desc += getString(R.string.color_temperature_title) + " : " + profile.getSettings()[0] + "K";
            } else {
                desc += getString(R.string.blue_light) + " : " + profile.getSettings()[0] + ", " +
                        getString(R.string.green_light) + " : " + profile.getSettings()[1];
            }
            return desc;
        }
    }

    private void initProfileCreatorViews() {
        creatorView = View.inflate(this, R.layout.bottom_sheet_create_profile, null);

        editText = creatorView.findViewById(R.id.profile_name_set);
        editTextError = creatorView.findViewById(R.id.profile_name_error);
        nlSwitch = creatorView.findViewById(R.id.profile_night_light_switch);
        settingTitle1 = creatorView.findViewById(R.id.profile_night_light_setting_title1);
        settingTitle2 = creatorView.findViewById(R.id.profile_night_light_setting_title2);
        cancel = creatorView.findViewById(R.id.button_cancel);
        modes = creatorView.findViewById(R.id.profile_night_light_setting_mode);
        settingParam1 = creatorView.findViewById(R.id.profile_night_light_setting_param1);
        settingParam2 = creatorView.findViewById(R.id.profile_night_light_setting_param2);
        ok = creatorView.findViewById(R.id.button_ok);

        modes.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                currentModeSelection = curProfile == null ? position : curProfile.getSettingMode();
                updateProfileCreatorParams(currentModeSelection, curProfile);
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        cancel.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                bottomSheetDialog.dismiss();
            }
        });

        ok.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                boolean retVal;
                if (curMode == MODE_CREATE) {
                    retVal = profilesManager.createProfile(nlSwitch.isChecked(),
                            editText.getEditableText().toString(),
                            modes.getSelectedItemPosition(),
                            modes.getSelectedItemId() == Constants.NL_SETTING_MODE_TEMP ? new int[]{settingParam1.getProgress() + 3000} : new int[]{settingParam1.getProgress(), settingParam2.getProgress()});
                } else {
                    retVal = profilesManager.updateProfile(curProfile.getName(),
                            nlSwitch.isChecked(),
                            editText.getEditableText().toString(),
                            modes.getSelectedItemPosition(),
                            modes.getSelectedItemId() == Constants.NL_SETTING_MODE_TEMP ? new int[]{settingParam1.getProgress() + 3000} : new int[]{settingParam1.getProgress(), settingParam2.getProgress()});
                }
                if (retVal) {
                    profiles = profilesManager.getProfilesList();
                    profilesAdapter.notifyDataSetChanged();
                    bottomSheetDialog.dismiss();
                } else editTextError.setVisibility(View.VISIBLE);
            }
        });

        if (curProfile != null) {
            nlSwitch.setChecked(curProfile.isSettingEnabled());
            editText.setText(curProfile.getName());
            modes.setSelection(curProfile.getSettingMode());
        }

        editTextError.setVisibility(View.GONE);
    }

    private void updateProfileCreatorParams(int mode, ProfilesManager.Profile profile) {
        if (mode == Constants.NL_SETTING_MODE_FILTER) {
            settingParam1.setEnabled(true);
            settingParam1.setMax(128);
            settingParam2.setEnabled(true);
            settingParam2.setMax(48);
            settingTitle1.setEnabled(true);
            settingTitle2.setEnabled(true);
            settingTitle1.setText(R.string.blue_light);
            settingTitle2.setText(R.string.green_light);
            if (profile != null) {
                settingParam1.setProgress(profile.getSettings()[0]);
                settingParam2.setProgress(profile.getSettings()[1]);
            } else {
                settingParam1.setProgress(PreferenceHelper.getInt(this, Constants.PREF_BLUE_INTENSITY, Constants.DEFAULT_BLUE_INTENSITY));
                settingParam2.setProgress(PreferenceHelper.getInt(this, Constants.PREF_GREEN_INTENSITY, Constants.DEFAULT_GREEN_INTENSITY));
            }
        } else {
            settingParam1.setEnabled(true);
            settingParam1.setMax(1500);
            settingParam2.setEnabled(false);
            settingTitle1.setEnabled(true);
            settingTitle2.setEnabled(false);
            settingTitle1.setText(R.string.color_temperature_title);
            settingTitle2.setText(R.string.profile_nl_setting_unavailable);
            if (profile != null) {
                settingParam1.setProgress(profile.getSettings()[0]-3000);
            } else {
                settingParam1.setProgress(PreferenceHelper.getInt(this, Constants.PREF_COLOR_TEMP, Constants.DEFAULT_COLOR_TEMP));
            }
        }
    }

    private void getOptionsView() {
        optionsView = View.inflate(this, R.layout.bottom_sheet_profile_options, null);

        optionsView.findViewById(R.id.apply).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (curProfile != null) curProfile.apply(ProfilesActivity.this);
                optionsDialog.dismiss();
            }
        });

        optionsView.findViewById(R.id.edit).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                curMode = MODE_EDIT;
                bottomSheetDialog = new BottomSheetDialog(ProfilesActivity.this, R.style.BottomSheetDialogDark);
                initProfileCreatorViews();
                bottomSheetDialog.setContentView(creatorView);
                bottomSheetDialog.setCancelable(false);
                bottomSheetDialog.show();
                optionsDialog.dismiss();
            }
        });

        optionsView.findViewById(R.id.delete).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                profilesManager.deleteProfile(curProfile.getName());
                profiles.remove(curProfile);
                curProfile = null;
                profilesAdapter.notifyDataSetChanged();
                optionsDialog.dismiss();
            }
        });
    }
}

File: app/src/main/java/com/corphish/nightlight/ProfilesActivity.kt
package com.corphish.nightlight

import android.content.Context
import android.content.Intent
import android.graphics.PorterDuff
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.corphish.nightlight.data.Constants
import com.corphish.nightlight.design.ThemeUtils
import com.corphish.nightlight.design.alert.BottomSheetAlertDialog
import com.corphish.nightlight.design.views.ProfileCreator
import com.corphish.nightlight.engine.ProfilesManager
import com.corphish.nightlight.helpers.PreferenceHelper
import com.google.android.material.bottomsheet.BottomSheetDialog
import kotlinx.android.synthetic.main.activity_profiles.*
import kotlinx.android.synthetic.main.content_profiles.*
import kotlinx.android.synthetic.main.layout_header.*

import java.util.Arrays

class ProfilesActivity : AppCompatActivity(), ProfilesManager.DataChangeListener {

    private lateinit var optionsDialog: BottomSheetDialog
    private lateinit var optionsView: View

    private var curProfile: ProfilesManager.Profile? = null

    private lateinit var profilesManager: ProfilesManager

    private lateinit var profilesAdapter: ProfilesAdapter

    private var profiles: MutableList<ProfilesManager.Profile>? = null

    private lateinit var context: Context

    private var midTemp = 0
    private var midKCALSum = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(ThemeUtils.getAppTheme(this))
        setContentView(R.layout.activity_profiles)

        context = this

        fab.setOnClickListener {
            ProfileCreator(this@ProfilesActivity, ProfileCreator.MODE_CREATE,
                    onFinishListener =  fun(status: Int) {
                        if (status == ProfileCreator.STATUS_SUCCESS) {
                            profilesManager.loadProfiles()
                            profiles = profilesManager.profilesList
                            profilesAdapter.notifyDataSetChanged()
                        }
                    }).show()
        }

        banner_title.text = getString(R.string.profile_title)
        banner_icon.setImageResource(R.drawable.ic_profiles_24dp)

        initProfilesManager()
        initViews()
        initMidValues()
    }

    private fun initMidValues() {
        midTemp = (PreferenceHelper.getInt(context, Constants.PREF_MAX_COLOR_TEMP, Constants.DEFAULT_MAX_COLOR_TEMP) +
                PreferenceHelper.getInt(context, Constants.PREF_MIN_COLOR_TEMP, Constants.DEFAULT_MIN_COLOR_TEMP))/2

        midKCALSum = (PreferenceHelper.getInt(context, Constants.PREF_MAX_RED_COLOR, Constants.DEFAULT_MAX_RED_COLOR) +
                PreferenceHelper.getInt(context, Constants.PREF_MIN_RED_COLOR, Constants.DEFAULT_MIN_RED_COLOR))/2
                +
                (PreferenceHelper.getInt(context, Constants.PREF_MAX_GREEN_COLOR, Constants.DEFAULT_MAX_GREEN_COLOR) +
                        PreferenceHelper.getInt(context, Constants.PREF_MIN_GREEN_COLOR, Constants.DEFAULT_MIN_GREEN_COLOR))/2
                +
                (PreferenceHelper.getInt(context, Constants.PREF_MAX_BLUE_COLOR, Constants.DEFAULT_MAX_BLUE_COLOR) +
                        PreferenceHelper.getInt(context, Constants.PREF_MIN_BLUE_COLOR, Constants.DEFAULT_MIN_BLUE_COLOR))/2
    }

    private fun initProfilesManager() {
        profilesManager = ProfilesManager(this)
        profilesManager.registerDataChangeListener(this)
        profilesManager.loadProfiles()
        profiles = profilesManager.profilesList
    }

    private fun initViews() {
        profilesAdapter = ProfilesAdapter()
        profilesAdapter.setProfiles(profiles)

        recyclerView.invalidateItemDecorations()
        recyclerView.layoutManager = GridLayoutManager(this, 4)
        recyclerView.adapter = profilesAdapter
        recyclerView.itemAnimator = androidx.recyclerview.widget.DefaultItemAnimator()
        recyclerView.isNestedScrollingEnabled = false
        recyclerView.setHasFixedSize(false)

        profilesAdapter.notifyDataSetChanged()
    }

    override fun onDataChanged(newDataSize: Int) {
        if (newDataSize < 1)
            emptyView.visibility = View.VISIBLE
        else
            emptyView.visibility = View.GONE
    }

    private inner class ProfilesAdapter : RecyclerView.Adapter<ProfilesActivity.ProfilesAdapter.CustomViewHolder>() {
        private var profiles: List<ProfilesManager.Profile>? = null

        internal fun setProfiles(profiles: MutableList<ProfilesManager.Profile>?) {
            this.profiles = profiles
        }

        inner class CustomViewHolder internal constructor(v: View) : RecyclerView.ViewHolder(v), View.OnClickListener {
            internal val icon: TextView = v.findViewById(R.id.profileIcon)
            internal val title: TextView = v.findViewById(R.id.profileTitle)

            init {
                v.setOnClickListener(this)
            }

            override fun onClick(v: View) {
                if (!intent.getBooleanExtra(Constants.TASKER_ERROR_STATUS, true)) {
                    showAlert(R.string.confirm, getString(R.string.tasker_confirm_selection, profiles!![adapterPosition].name), View.OnClickListener { returnBack(profiles!![adapterPosition].name) })
                } else {
                    curProfile = profiles!![adapterPosition]
                    optionsDialog = BottomSheetDialog(this@ProfilesActivity, ThemeUtils.getBottomSheetTheme(context))
                    getOptionsView(curProfile!!)
                    optionsDialog.setContentView(optionsView)
                    optionsDialog.show()
                }
            }
        }


        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomViewHolder {
            val itemView = LayoutInflater.from(parent.context)
                    .inflate(R.layout.layout_profile_item, parent, false)

            return CustomViewHolder(itemView)
        }

        override fun onBindViewHolder(holder: CustomViewHolder, position: Int) {
            val profile = profiles!![position]
            val profileValue = profile.settings.sum()
            val profileIntensity =
                    if (profile.settingMode == Constants.NL_SETTING_MODE_TEMP)
                        if (profileValue > midTemp) Constants.INTENSITY_TYPE_MINIMUM else Constants.INTENSITY_TYPE_MAXIMUM
                    else
                        if (profileValue > midKCALSum) Constants.INTENSITY_TYPE_MINIMUM else Constants.INTENSITY_TYPE_MAXIMUM

            holder.icon.text = if (profile.name.isNotEmpty()) "${profile.name.toUpperCase()[0]}" else ""
            setIconBackground(holder.icon, ThemeUtils.getNLStatusIconBackground(context, profile.isSettingEnabled, profileIntensity))
            holder.icon.setTextColor(ThemeUtils.getNLStatusIconForeground(context, profile.isSettingEnabled, profileIntensity))

            holder.title.text = profile.name
        }

        private fun setIconBackground(textView: TextView, color: Int) {
            val drawable = resources.getDrawable(R.drawable.circle)
            drawable.setColorFilter(color, PorterDuff.Mode.SRC)

            textView.background = drawable
        }

        override fun getItemCount(): Int {
            return profiles!!.size
        }
    }



    private fun getOptionsView(profile: ProfilesManager.Profile?) {
        optionsView = View.inflate(this, R.layout.bottom_sheet_profile_options, null)

        val selectedProfileName = optionsView.findViewById<TextView>(R.id.selectedProfileName)
        val selectedProfileInfo = optionsView.findViewById<TextView>(R.id.selectedProfileInfo)
        val apply = optionsView.findViewById<View>(R.id.apply)
        val edit = optionsView.findViewById<View>(R.id.edit)
        val delete = optionsView.findViewById<View>(R.id.delete)

        selectedProfileName.text = profile?.name
        selectedProfileInfo.text = "${arrayOf(getString(R.string.color_temperature_title), "RGB")[profile?.settingMode!!]}: ${Arrays.toString(profile?.settings!!)}"

        apply.setOnClickListener {
            if (curProfile != null) {
                curProfile!!.apply(this@ProfilesActivity)
                PreferenceHelper.putInt(context, Constants.PREF_CUR_APPLY_TYPE, Constants.APPLY_TYPE_PROFILE)
                PreferenceHelper.putBoolean(context, Constants.PREF_CUR_APPLY_EN, curProfile!!.isSettingEnabled)
                PreferenceHelper.putInt(context, Constants.PREF_CUR_PROF_MODE, curProfile!!.settingMode)
                PreferenceHelper.putString(context, Constants.PREF_CUR_PROF_VAL, Arrays.toString(curProfile!!.settings))
            }
            optionsDialog.dismiss()
        }

        edit.setOnClickListener {
            ProfileCreator(this@ProfilesActivity, ProfileCreator.MODE_EDIT, profile,
                    onFinishListener =  fun(status: Int) {
                        if (status == ProfileCreator.STATUS_SUCCESS) {
                            profiles = profilesManager.profilesList
                            profilesAdapter.notifyDataSetChanged()
                        }
                    }).show()
            optionsDialog.dismiss()
        }

        delete.setOnClickListener {
            showAlert(R.string.delete, getString(R.string.delete_details, curProfile!!.name), View.OnClickListener {
                profilesManager.deleteProfile(curProfile!!)
                val prof = curProfile
                profiles!!.remove(prof)
                curProfile = null
                profilesAdapter.notifyDataSetChanged()
            })
            optionsDialog.dismiss()
        }
    }

    fun returnBack(name: String?) {
        if (!intent.getBooleanExtra(Constants.TASKER_ERROR_STATUS, true)) {
            val bundle = Bundle()
            bundle.putString(Constants.TASKER_SETTING, name)
            val intent = Intent()
            intent.putExtra("com.twofortyfouram.locale.intent.extra.BLURB", name)
            intent.putExtra("com.twofortyfouram.locale.intent.extra.BUNDLE", bundle)
            setResult(RESULT_OK, intent)
            finish()
        }
    }

    private fun showAlert(title: Int, message: String, positiveOnClickListener: View.OnClickListener) {
        val bottomSheetAlertDialog = BottomSheetAlertDialog(this)
        bottomSheetAlertDialog.setTitle(title)
        bottomSheetAlertDialog.setMessage(message)
        bottomSheetAlertDialog.setPositiveButton(android.R.string.ok, positiveOnClickListener)
        bottomSheetAlertDialog.setNegativeButton(android.R.string.cancel, View.OnClickListener { })
        bottomSheetAlertDialog.show()
    }
}
